external/devlib/devlib/trace/dmesg.py

```diff
@@ -0,0 +1,187 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import division
+import re
+from itertools import takewhile
+from datetime import timedelta
+
+from devlib.trace import TraceCollector
+
+
+class KernelLogEntry(object):
+    """
+    Entry of the kernel ring buffer.
+
+    :param facility: facility the entry comes from
+    :type facility: str
+
+    :param level: log level
+    :type level: str
+
+    :param timestamp: Timestamp of the entry
+    :type timestamp: datetime.timedelta
+
+    :param msg: Content of the entry
+    :type msg: str
+    """
+
+    _TIMESTAMP_MSG_REGEX = re.compile(r'\[(.*?)\] (.*)')
+
+    def __init__(self, facility, level, timestamp, msg):
+        self.facility = facility
+        self.level = level
+        self.timestamp = timestamp
+        self.msg = msg
+
+    @classmethod
+    def from_str(cls, line):
+        """
+        Parses a "dmesg --decode" output line, formatted as following:
+        kern  :err   : [3618282.310743] nouveau 0000:01:00.0: systemd-logind[988]: nv50cal_space: -16
+
+        Or the more basic output:
+        [3618282.310743] nouveau 0000:01:00.0: systemd-logind[988]: nv50cal_space: -16
+
+        """
+
+        def parse_timestamp_msg(line):
+            match = cls._TIMESTAMP_MSG_REGEX.match(line.strip())
+            if not match:
+                raise ValueError('dmesg entry format not recognized: {}'.format(line))
+            timestamp, msg = match.groups()
+            return timestamp, msg
+
+        # If we can parse the timestamp directly, that is a basic line
+        try:
+            timestamp, msg = parse_timestamp_msg(line)
+        except ValueError:
+            facility, level, remainder = line.split(':', 2)
+            timestamp, msg = parse_timestamp_msg(remainder)
+            facility = facility.strip()
+            level = level.strip()
+        else:
+            facility = None
+            level = None
+
+        return cls(
+            facility=facility,
+            level=level,
+            timestamp=timedelta(seconds=float(timestamp.strip())),
+            msg=msg.strip(),
+        )
+
+    def __str__(self):
+        if self.facility and self.level:
+            prefix = '{facility}:{level}:'.format(
+                facility=self.facility,
+                level=self.level,
+            )
+        else:
+            prefix = ''
+
+        return '{prefix}[{timestamp}] {msg}'.format(
+            timestamp=self.timestamp.total_seconds(),
+            msg=self.msg,
+            prefix=prefix,
+        )
+
+
+class DmesgCollector(TraceCollector):
```

Comment:
This inheritance relationship doesn't sound right to me. From naming point of view at least. What facilities in the parent class you care about here?

If it is to auto start/stop then it seems to me TraceCollector and DmesgCollector should have a common parent class. 

e.g
```
                   BufferCollector
                         |
          +------------- +----------------+
          |                               |
      TraceCollector                  DmesgCollector
```

Reply:
There are already a number of other collectors in devlib like `LogcatCollector`, which already inherit from `TraceCollector`. That hierarchy of names probably needs to be preserved for backward compatibilities, but if you think it's worth a rename, having aliases is possible if devlib maintainers are happy with that.

Reply:
> What facilities in the parent class you care about here?

Mostly the context manager (`__enter__`/`__exit__`) and the "promise on the API" (all collectors can be expected to behave roughly the same way without looking at each specific case). I initially tried to implement it as an instrument, but it was not suited at all. `TraceCollector` seemed to have a number of derived classes doing more or less the same thing as I do.